cluster/assemblestate/assemblestate.go

```diff
@@ -0,0 +1,880 @@
+	// the RDTs that they are is associated with.
+	fingerprints map[DeviceToken]Fingerprint
+
+	// addresses keeps track of which address we can reach each device at.
+	// Presence in this map does not imply trust. Additionally, a device can be
+	// trusted before we have an address.
+	addresses map[Fingerprint]string
+
+	// discovered keeps track of which addresses we've already discovered. We
+	// won't re-send auth messages to these addresses.
+	discovered map[string]bool
+
+	// devices keeps track of device identities. Additionally, it helps manage
+	// the events that trigger responses to device queries and the events that
+	// result in us sending our own queries.
+	devices DeviceQueryTracker
+
+	// selector keeps track of our routes and decides the strategy for
+	// publishing routes to our peers.
+	selector RouteSelector
+}
+
+// AssembleSession provides a method for serializing our current state of
+// assembly to JSON.
+type AssembleSession struct {
+	Initiated  time.Time              `json:"initiated"`
+	Trusted    map[string]Peer        `json:"trusted"`
+	Addresses  map[string]string      `json:"addresses"`
+	Discovered []string               `json:"discovered"`
+	Routes     Routes                 `json:"routes"`
+	Devices    DeviceQueryTrackerData `json:"devices"`
+}
+
+func (as *AssembleState) export() AssembleSession {
+	trusted := make(map[string]Peer, len(as.trusted))
+	for fp, p := range as.trusted {
+		trusted[base64.StdEncoding.EncodeToString(fp[:])] = p
+	}
+
+	addresses := make(map[string]string, len(as.addresses))
+	for fp, addr := range as.addresses {
+		addresses[base64.StdEncoding.EncodeToString(fp[:])] = addr
+	}
+
+	discovered := make([]string, 0, len(as.discovered))
+	for addr := range as.discovered {
+		discovered = append(discovered, addr)
+	}
+
+	return AssembleSession{
+		Initiated:  as.initiated,
+		Trusted:    trusted,
+		Addresses:  addresses,
+		Discovered: discovered,
+		Routes:     as.selector.Routes(),
+		Devices:    as.devices.Export(),
+	}
+}
+
+type Peer struct {
+	RDT  DeviceToken `json:"rdt"`
+	Cert []byte      `json:"cert"`
+}
+
+const AssembleSessionLength = time.Hour
+
+// NewAssembleState creates a new [AssembleState] from the given configuration
+// and session data.
+func NewAssembleState(
+	config AssembleConfig,
+	session AssembleSession,
+	selector func(self DeviceToken, identified func(DeviceToken) bool) (RouteSelector, error),
+	log logger.Logger,
+	commit func(AssembleSession),
+) (*AssembleState, error) {
+	// default clock to time.Now if not provided
+	if config.Clock == nil {
+		config.Clock = time.Now
+	}
+
+	if !session.Initiated.IsZero() && config.Clock().Sub(session.Initiated) > AssembleSessionLength {
+		return nil, errors.New("cannot resume an assembly session that began more than an hour ago")
+	}
+
+	// validate the given session parse it into more useful data structures
+	validated, err := validateSession(session)
```

Comment:
would it be clearer if we copied more things into validated and used validated below everywhere instead of a mix of validated and session?